mono/metadata/marshal.c

```diff
@@ -4938,183 +4958,131 @@ mono_marshal_set_last_error_windows (int error)
 		mono_error_set_argument (error, "length", "start_index + length > array length");
-		mono_error_set_pending_exception (error);
-		return;
+		return 0;
 	}
 
-	element_size = mono_array_element_size (src->obj.vtable->klass);
+	gsize element_size = mono_array_element_size (klass);
 
-	/* no references should be involved */
-	source_addr = mono_array_addr_with_size_fast (src, element_size, start_index);
+	// Handle generic arrays.
+	// FIXME Given this is required for generic arrays, is the support for non-generic arrays worthwhile?
+	if (!*managed_addr)
+		*managed_addr = mono_array_handle_pin_with_size (managed, element_size, start_index, gchandle);
 
-	memcpy (dest, source_addr, length * element_size);
+	return element_size * length;
 }
 
 void
-ves_icall_System_Runtime_InteropServices_Marshal_copy_from_unmanaged (gpointer src, gint32 start_index,
-								      MonoArray *dest, gint32 length)
+ves_icall_System_Runtime_InteropServices_Marshal_copy_to_unmanaged (MonoArrayHandle src, gint32 start_index,
+		gpointer dest, gint32 length, gconstpointer managed_source_addr, MonoError *error)
 {
-	int element_size;
-	void *dest_addr;
-	ERROR_DECL (error);
-
-	MONO_CHECK_ARG_NULL (src,);
-	MONO_CHECK_ARG_NULL (dest,);
-
-	if (m_class_get_rank (mono_object_class (&dest->obj)) != 1) {
-		mono_error_set_argument (error, "array", "array is multi-dimensional");
-		mono_error_set_pending_exception (error);
-		return;
-	}
-	if (start_index < 0) {
-		mono_error_set_argument (error, "startIndex", "Must be >= 0");
-		mono_error_set_pending_exception (error);
-		return;
-	}
-	if (length < 0) {
-		mono_error_set_argument (error, "length", "Must be >= 0");
-		mono_error_set_pending_exception (error);
-		return;
-	}
-	if (start_index + length > mono_array_length (dest)) {
-		ERROR_DECL (error);
-		mono_error_set_argument (error, "length", "start_index + length > array length");
-		mono_error_set_pending_exception (error);
-		return;
-	}
-	element_size = mono_array_element_size (dest->obj.vtable->klass);
-	  
-	/* no references should be involved */
-	dest_addr = mono_array_addr_with_size_fast (dest, element_size, start_index);
+	guint32 gchandle = 0;
+	gsize const bytes = copy_managed_common (src, dest, start_index, length, (gpointer*)&managed_source_addr, &gchandle, error);
+	if (bytes && managed_source_addr && dest && length > 0 && start_index >= 0 && is_ok (error))
+		memcpy (dest, managed_source_addr, bytes); // no references should be involved
+	mono_gchandle_free (gchandle);
+}
 
-	memcpy (dest_addr, src, length * element_size);
+void
+ves_icall_System_Runtime_InteropServices_Marshal_copy_from_unmanaged (gconstpointer src, gint32 start_index,
+		MonoArrayHandle dest, gint32 length, gpointer managed_dest_addr, MonoError *error)
+{
+	guint32 gchandle = 0;
+	gsize const bytes = copy_managed_common (dest, src, start_index, length, &managed_dest_addr, &gchandle, error);
+	if (bytes && managed_dest_addr && src && length > 0 && start_index >= 0 && is_ok (error))
```

Comment:
Some/most of this if is redundant but probably ok. i.e. mainly `bytes` needs to be checked, or if `is_ok`.